Handle Masking and substitution for NaN & FILLVAL Values (#92)
swxsoc/swxsoc: docs/user-guide/fillval_and_masks.rst

.. _fillval_and_masks:

**********************************

> FILLVAL, NaN, and Mask Round-trips

Reviewer: ```suggestion
Tagging missing or bad data
```

**********************************

Overview
========

> ``swxsoc`` round-trips missing data between in-memory containers and CDF files using the ISTP FILLVAL convention.

Reviewer: ```suggestion
``swxsoc`` supports tagging missing or bad data with our in-memory containers and seamlessly converts these tags values in CDF files using the ISTP FILLVAL convention.
```

Masks are the canonical in-memory representation of missing data.

> For convenience, floating-point NaN values are also treated as missing on the write path.

Reviewer: ```suggestion
For convenience, floating-point NaN values are also treated as missing.
```

> Integer and string dtypes are never promoted to floats; missing positions are tracked exclusively through a boolean mask.

Reviewer: ```suggestion
For integer and string dtypes missing positions are tracked exclusively through a boolean mask (even though `~numpy.nan` are internally floats).
```

> Per-dtype behavior

Reviewer: ```suggestion
Missing values across data types
```

==================

.. list-table::
   :header-rows: 1
   :widths: 15 30 30 25

   * - Dtype
     - In-memory fill source
     - On disk
     - On read result
   * - Float
     - ``NaN`` or ``.mask`` bit set

>      - ``FILLVAL`` (e.g. ``-1.0e31``)

Reviewer: What if the user wants to use a custom FILLVAL? This may occur if the instrument explictely outputs a specific value when it does not have a measurements (e.g. time-out).

Author: The wording here makes it sound like you should not define your own FILVAL. https://spdf.gsfc.nasa.gov/istp_guide/vattributes.html#FILLVAL

   # After write/read round-trip, ``loaded`` is a Masked Quantity with
   # mask=[False, True, False] and underlying value [1.0, nan, 3.0].

Integers
========

Integer dtypes are never promoted to float.
Missing positions must be carried as a boolean ``.mask`` on an :class:`~astropy.nddata.NDData` (or :class:`~ndcube.NDCube`)::

   import numpy as np
   from astropy.nddata import NDData

   nd = NDData(
       data=np.array([10, 20, 30, 40], dtype=np.int16),
       mask=np.array([False, True, False, True]),
   )

On disk, masked positions are written as the variable's ``FILLVAL`` sentinel (chosen from the schema).
On read, the dtype is preserved and the mask is restored.
If a sentinel value appears in the on-disk data without an explicit mask, it is treated as fill on read.

Strings
=======

String round-trips are intentionally asymmetric:

* On write, any position equal to the literal bytes ``b"nan"`` or ``b"NaN"`` is treated as fill (this is how numpy coerces ``np.nan`` into ``S``/``U`` arrays) and is emitted to the CDF as a single space ``b" "``.
* On read, only the spec sentinel ``b" "`` is mapped to a mask bit.
  The literal string ``"nan"`` is never reinterpreted on read.

> A consequence: a legitimate string value of ``"nan"`` will be coerced to fill on write.

Reviewer: Above these are labelled as bytes but here it is a string but then the next sentence refers to the bytes.

If you need to preserve the literal four-byte string ``"nan"``, do not use it as a data value.

Time / Epoch
============

> Time columns use astropy's native masking on :class:`~astropy.time.Time`::

Reviewer: what happens if the time is masked but the measurement is not?

Author: Time columns and measurements go through independent paths. The two masks are decoupled. The time column carries "time unknown at i=2"; the measurement column carries a real value at i=2 with no indication that the timestamp is missing. The user (or downstream code) is responsible for propagating a time mask into data columns if that's the desired semantics.

   from astropy.time import Time
   import numpy as np

   t = Time(np.arange(5), format="unix")
   t[2] = np.ma.masked

When a mask is present, the writer bypasses the datetime conversion path and writes raw ``TT2000`` nanoseconds (``int64``) so the ISTP sentinel ``-9223372036854775808`` is preserved exactly.
On read, the time column comes back as a ``Time`` whose ``.mask`` is set at sentinel positions; ``.masked`` is ``True``.

Working with masks
==================

> After loading a CDF, you can inspect masks uniformly across container types::

Reviewer: ```suggestion
After loading a CDF, you can inspect missing values through the mask uniformly across types::
```
